dandiapi/api/services/publish/__init__.py

```diff
@@ -0,0 +1,136 @@
+
+
+def _publish_dandiset(dandiset_id: int) -> None:
+    """
+    Publish a dandiset.
+
+    Calling `_lock_dandiset_for_publishing()` is a precondition for calling this function.
+    """
+    old_version: Version = (
+        Dandiset.objects.get(id=dandiset_id).versions.select_for_update().get(version='draft')
+    )
+
+    with transaction.atomic():
+        if old_version.status != Version.Status.PUBLISHING:
+            raise DandisetNotLocked(
+                'Dandiset must be in PUBLISHING state. Call `_lock_dandiset_for_publishing()` '
+                'before this function.'
+            )
+
+        new_version: Version = old_version.publish_version
+        new_version.save()
+
+        # Bulk create the join table rows to optimize linking assets to new_version
+        AssetVersions = Version.assets.through
+
+        # Add a new many-to-many association directly to any already published assets
+        already_published_assets = old_version.assets.filter(published=True)
+        AssetVersions.objects.bulk_create(
+            [
+                AssetVersions(asset_id=asset['id'], version_id=new_version.id)
+                for asset in already_published_assets.values('id')
+            ]
+        )
+
+        # Publish any draft assets
+        draft_assets = old_version.assets.filter(published=False).all()
+        for draft_asset in draft_assets:
+            draft_asset.publish()
+
+        Asset.objects.bulk_update(draft_assets, ['metadata', 'published'])
+
+        AssetVersions.objects.bulk_create(
+            [AssetVersions(asset_id=asset.id, version_id=new_version.id) for asset in draft_assets]
+        )
+
+        # Save again to recompute metadata, specifically assetsSummary
+        new_version.save()
+
+        # Add asset paths with new version
+        add_version_asset_paths(version=new_version)
+
+        # Set the version of the draft to PUBLISHED so that it cannot be published again without
+        # being modified and revalidated
+        old_version.status = Version.Status.PUBLISHED
+        old_version.save()
+
+        # Write updated manifest files and create DOI after
+        # published version has been committed to DB.
+        transaction.on_commit(lambda: write_manifest_files.delay(new_version.id))
+
+        def _create_doi(version_id: int):
+            version = Version.objects.get(id=version_id)
+            version.doi = doi.create_doi(version)
+            version.save()
+
+        transaction.on_commit(lambda: _create_doi(new_version.id))
+
+
+def publish_dandiset(*, user: User, dandiset: Dandiset) -> None:
+    from dandiapi.api.tasks import publish_dandiset_task
+
+    if (
+        not user.has_perm('owner', dandiset)
+        or dandiset.embargo_status != Dandiset.EmbargoStatus.OPEN
+    ):
+        raise NotAllowed()
+    if dandiset.zarr_archives.exists() or dandiset.embargoed_zarr_archives.exists():
+        raise NotAllowed('Cannot publish dandisets which contain zarrs.')
```

Comment:
It seems this exact code block exists at the beginning of `_lock_dandiset_for_publishing`, which immediately follows here. So it seems this check is redundant? I'd understand if it had to stay to preserve order of operations, but it seems that wouldn't be impacted by its removal.

Reply:
[`ac75bf5` (#1363)](https://github.com/dandi/dandi-archive/pull/1363/commits/ac75bf5dd61958debc0ac32d50e4be81b39be850)